// @pinefile/pine/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.run = exports.shell = exports.Logger = exports.createLogger = exports.getConfig = exports.configure = exports.color = exports.getArgs = exports.filterArgs = exports.api = void 0;
const cli_1 = require("./cli");
const file_1 = require("./file");
const task_1 = require("./task");
const runner_1 = require("./runner");
// api functions
const api = { runCLI: cli_1.runCLI, resolveTask: task_1.resolveTask, runTask: task_1.runTask, createRunner: runner_1.createRunner, loadPineFile: file_1.loadPineFile };
exports.api = api;
// core functions
var args_1 = require("./args");
Object.defineProperty(exports, "filterArgs", { enumerable: true, get: function () { return args_1.filterArgs; } });
Object.defineProperty(exports, "getArgs", { enumerable: true, get: function () { return args_1.getArgs; } });
var color_1 = require("./color");
Object.defineProperty(exports, "color", { enumerable: true, get: function () { return color_1.color; } });
var config_1 = require("./config");
Object.defineProperty(exports, "configure", { enumerable: true, get: function () { return config_1.configure; } });
Object.defineProperty(exports, "getConfig", { enumerable: true, get: function () { return config_1.getConfig; } });
var logger_1 = require("./logger");
Object.defineProperty(exports, "createLogger", { enumerable: true, get: function () { return logger_1.createLogger; } });
Object.defineProperty(exports, "Logger", { enumerable: true, get: function () { return logger_1.Logger; } });
// plugin functions.
__exportStar(require("./plugins/log"), exports);
var shell_1 = require("./plugins/shell");
Object.defineProperty(exports, "shell", { enumerable: true, get: function () { return shell_1.shell; } });
Object.defineProperty(exports, "run", { enumerable: true, get: function () { return shell_1.run; } });
__exportStar(require("./plugins/task"), exports);
